projects/components/src/download-file/service/file-download.service.ts

```diff
@@ -44,6 +47,13 @@ export class FileDownloadService {
 
     // Convert values into strings
     const values$ = config.dataSource.pipe(
+      // All rows are not guaranteed to have all keys, so add all keys to all rows
+      map(data => {
+        const allKeys = getAllDataKeys(data);
+        const allKeysDict = allKeys.reduce((acc, key) => ({ ...acc, [key]: undefined }), {});
+
+        return data.map(row => ({ ...allKeysDict, ...row }));
```

Comment:
Because we're using object.values to iterate this below, we're still relying here on insertion order being consistent, which it won't be if we're spreading like this (also IIRC there's some nuances to the iteration order of object key/values, so best not to rely on it). suggest we iterate on keys to access values rather than using them as-is to overwrite empty values like this.

Reply:
Are you suggesting we do this?:
```ts
Object.fromEntries(allKeys.map((key) => [key, row?.[key] ?? undefined]));
```

Reply:
Are you suggesting we do this instead?:
```ts
data.map((row) => {
  const newRow = Object.fromEntries(allKeys.map((key) => [key, row?.[key] ?? undefined]));
  return newRow;
})
```

Reply:
there's multiple approaches here, but I probably would avoid mapping it into an object and back out again. Not only is it extra allocations, but more important - you're relying on semantics (both of .fromEntries and of .values). Instead, map allKeys to the stringify calls.

Reply:
Updated.